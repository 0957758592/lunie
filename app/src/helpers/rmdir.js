const fs = require('fs')

// under windows removing folders results in several errors EPERM, ENOTEMPTY
// adding the timeout solves these issues
async function deleteFolderRecursive(path) {
  if (fs.existsSync(path)) {
    fs.readdirSync(path).forEach(async function(file, index){
      var curPath = path + "/" + file;
      if (fs.lstatSync(curPath).isDirectory()) { // recurse
        await deleteFolderRecursive(curPath)
      } else { // delete file
        fs.unlinkSync(curPath);
      }
    });
    await sleep(10)
    fs.rmdirSync(path);
  }
}

function sleep (ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

module.exports = deleteFolderRecursive