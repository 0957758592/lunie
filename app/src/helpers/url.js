export function getURLParams(window) {
  const queries = window.location.search.slice(1).split(`&`)
  const parameters = queries.reduce((config, current) => {
    const [name, value] = current.split(`=`)
    if ([`stargate`, `rpc`, `experimental`].includes(name)) {
      return {
        ...config,
        [name]: value
      }
    }
    return config
  }, {})

  if (process.env.NODE_ENV === `production` && (parameters.stargate || parameters.rpc)) {
    alert(`The ability to set the remote stargate and full node was removed in production to improve security.`)
  }

  return parameters
}
