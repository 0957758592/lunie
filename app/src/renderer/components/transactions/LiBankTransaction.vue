<template>
  <li-transaction
    :color="`#ED553B`"
    :time="transaction.time"
    :block="transaction.height"
  >
    <template v-if="sent">
      <div slot="caption">
        Sent&nbsp;<b>{{ atoms(coins.amount) }}</b><span>&nbsp;{{ coins.denom.toUpperCase() }}</span>
      </div>
      <span
        slot="details"
      >
        <template
          v-if="sentSelf"
        >
          To yourself!
        </template><template
          v-else
        >
          To {{ receiver }}
        </template>
      </span>
    </template><template v-else>
      <div slot="caption">
        Received&nbsp;<b>{{ atoms(coins.amount) }}</b><span>&nbsp;{{ coins.denom.toUpperCase() }}</span>
      </div>
      <span slot="details">From {{ sender }}</span>
    </template>
  </li-transaction>
</template>

<script>
import LiTransaction from "./LiTransaction"
import { atoms } from "../../scripts/num.js"

export default {
  name: `li-bank-transaction`,
  components: { LiTransaction },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    address: {
      type: String,
      default: null
    }
  },
  data: () => ({
    atoms
  }),
  computed: {
    tx() {
      return this.transaction.tx.value.msg[0].value
    },
    // TODO: sum relevant inputs/outputs
    sentSelf() {
      return this.tx.from_address === this.tx.to_address
    },
    sent() {
      return this.tx.from_address === this.address
    },
    sender() {
      return this.tx.from_address
    },
    coins() {
      return this.tx.amount[0]
    },
    receiver() {
      return this.tx.to_address
    }
  }
}
</script>

<style>
.li-tx {
  display: flex;
  font-size: var(--sm);
}

.li-tx .tx-icon {
  padding: 0 0.5rem;
  background: var(--app-fg);
  display: flex;
  align-items: center;
  justify-content: center;
}

.li-tx .tx-container {
  flex-direction: column;
  flex-wrap: nowrap;
  padding: 0.5rem 0;
  margin: 0.5rem 0;
  display: flex;
  width: 100%;
  min-width: 0;
}

.li-tx .tx-element {
  padding: 0 2rem 0 1.5rem;
  line-height: 1.5rem;
}

.li-tx .tx-coin .value {
  flex: 0 0 100%;
  font-size: var(--sm);
  color: var(--dim);
}

.li-tx .tx-coin .value::before {
  content: "";
  display: inline;
}

.li-tx .tx-coin .key {
  font-weight: 500;
  font-size: var(--m);
}

.li-tx .tx-coin .value,
.li-tx .tx-coin .key {
  line-height: 1.5rem;
}

.li-tx .tx-address {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  color: var(--dim);
  font-size: var(--sm);
}

.li-tx.li-tx-sent .tx-coin .value::before {
  content: "-";
}

.li-tx.li-tx-received .tx-icon {
  background: var(--app-fg);
}

.li-tx.li-tx-received .tx-coin .value {
  color: var(--success);
}

.li-tx.li-tx-received .tx-coin .value::before {
  content: "+";
}

.li-tx:hover {
  cursor: pointer;
}

@media screen and (min-width: 700px) {
  .li-tx {
    font-size: 0.875rem;
  }

  .li-tx .tx-container {
    flex-direction: row;
  }

  .li-tx .tx-container .tx-coins {
    flex: 0 0 9rem;
    padding: 0;
    min-width: 0;
  }

  .li-tx .tx-container .tx-coins .tx-coin {
    padding: 0 1.5rem 0;
  }

  .li-tx .tx-container .tx-coins .tx-coin .key {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
}
</style>
