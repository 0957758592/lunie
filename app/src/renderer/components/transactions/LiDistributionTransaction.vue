<template>
  <li-transaction
    :color="`#F2B134`"
    :time="transaction.time"
    :block="transaction.height"
  >
    <template v-if="txType === `cosmos-sdk/MsgWithdrawDelegationReward`">
      <div slot="caption">
        Withdraw rewards
        <template v-if="fees">
          &nbsp;<b>{{ totalFeesOnly }}&nbsp;</b>
          <span>{{ feeDenom }}</span>
        </template>
      </div>
      <div slot="details">
        From&nbsp;<router-link :to="url + '/' + tx.validator_address">
          {{ moniker(tx.validator_address) }}
        </router-link>
      </div>
    </template>
    <template v-else-if="txType === `cosmos-sdk/MsgSetWithdrawAddress`">
      <div slot="caption">
        Update withdraw address
        <template v-if="fees">
          &nbsp;<b>{{ totalFeesOnly }}&nbsp;</b>
          <span>{{ feeDenom }}</span>
        </template>
      </div>
      <div slot="details">
        To {{ tx.withdraw_address }}
      </div>
    </template>
    <template
      v-else-if="txType === `cosmos-sdk/MsgWithdrawValidatorCommission`"
    >
      <div slot="caption">
        Withdraw validator commission
        <template v-if="fees">
          &nbsp;<b>{{ totalFeesOnly }}&nbsp;</b>
          <span>{{ feeDenom }}</span>
        </template>
      </div>
      <div slot="details">
        From&nbsp;<router-link :to="url + '/' + tx.validator_address">
          {{ moniker(tx.validator_address) }}
        </router-link>
      </div>
    </template>
  </li-transaction>
</template>

<script>
import LiTransaction from "./LiTransaction"
import { pretty, atoms, full } from "../../scripts/num.js"

export default {
  name: `li-distribution-transaction`,
  components: { LiTransaction },
  props: {
    transaction: {
      type: Object,
      required: true
    },
    fees: {
      type: Object,
      default: null
    },
    url: {
      type: String,
      required: true
    },
    bondingDenom: {
      type: String,
      required: true
    },
    txType: {
      type: String,
      required: true
    },
    validators: {
      type: Array,
      required: true
    }
  },
  data: () => ({
    atoms,
    full,
    pretty
  }),
  computed: {
    tx() {
      return this.transaction.tx.value.msg[0].value
    },
    totalFeesOnly({ fees, full, atoms, bondingDenom } = this) {
      if (fees && fees[bondingDenom]) {
        return full(atoms(fees[bondingDenom]))
      }
      return ``
    },
    feeDenom({ fees } = this) {
      if (fees) {
        const feeDenoms = Object.keys(fees)
        return `${feeDenoms[0]}s`
      }
      return ``
    }
  },
  methods: {
    moniker(validatorAddr) {
      const validator = this.validators.find(
        c => c.operator_address === validatorAddr
      )
      return validator ? validator.description.moniker : validatorAddr
    }
  }
}
</script>
