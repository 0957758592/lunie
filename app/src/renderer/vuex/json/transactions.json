{
  "dri3bdow": {
    "gas": 114,
    "fee": { "denom": "test", "amount": 125 },
    "inputs": [
      {
        "address": "Buffer",
        "coins": [
          { "denom": "test", "amount": 1000000 },
          { "denom": "atom", "amount": 4932 }
        ],
        "sequence": 1,
        "signature": "Buffer",
        "pubkey": "PubKey"
       }
    ],
    "outputs": [
      {
        "address": "Buffer",
        "coins": [
          { "denom": "test", "amount": 1000000 },
          { "denom": "atom", "amount": 40392 }
        ]
      }
    ]
  },
  "r03jf04n": {
    "gas": 343,
    "fee": { "denom": "test", "amount": 231 },
    "inputs": [
      {
        "address": "Buffer",
        "coins": [
          { "denom": "test", "amount": 1000000 },
          { "denom": "atom", "amount": 1234 }
        ],
        "sequence": 1,
        "signature": "Buffer",
        "pubkey": "PubKey"
       }
    ],
    "outputs": [
      {
        "address": "Buffer",
        "coins": [
          { "denom": "test", "amount": 1000000 },
          { "denom": "atom", "amount": 1234 }
        ]
      }
    ]
  },
  "boe9rbdi": {
    "gas": 249,
    "fee": { "denom": "test", "amount": 294 },
    "inputs": [
      {
        "address": "Buffer",
        "coins": [
          { "denom": "test", "amount": 1000000 },
          { "denom": "atom", "amount": 1234 }
        ],
        "sequence": 1,
        "signature": "Buffer",
        "pubkey": "PubKey"
       }
    ],
    "outputs": [
      {
        "address": "Buffer",
        "coins": [
          { "denom": "test", "amount": 1000000 },
          { "denom": "atom", "amount": 1234 }
        ]
      }
    ]
  }
}
