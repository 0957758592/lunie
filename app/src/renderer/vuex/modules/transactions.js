import { uniqBy } from "lodash"
import * as Sentry from "@sentry/browser"
import Vue from "vue"

export default ({ node }) => {
  const emptyState = {
    loading: false,
    loaded: false,
    error: null,
    bank: [], // {height, result: { gas, tags }, tx: { type, value: { fee: { amount: [{denom, amount}], gas}, msg: {type, inputs, outputs}}, signatures} }}
    staking: [],
    governance: [],
    distribution: []
  }
  const state = JSON.parse(JSON.stringify(emptyState))

  const mutations = {
    setBankTxs(state, txs) {
      Vue.set(state, `bank`, txs)
    },
    setStakingTxs(state, txs) {
      Vue.set(state, `staking`, txs)
    },
    setGovernanceTxs(state, txs) {
      Vue.set(state, `governance`, txs)
    },
    setDistributionTxs(state, txs) {
      Vue.set(state, `distribution`, txs)
    },
    setHistoryLoading(state, loading) {
      Vue.set(state, `loading`, loading)
    },
    setTransactionTime(state, { blockHeight, time }) {
      [`staking`, `bank`, `governance`, `distribution`]
        .forEach(category => {
          state[category].forEach(tx => {
            if (tx.height === blockHeight && time) {
              // time seems to be an ISO string, but we are expecting a Number type
              time = new Date(time).getTime()
              Vue.set(tx, `time`, time)
            }
          })
        })
    }
  }

  const actions = {
    resetSessionData({ rootState }) {
      // clear previous account state
      rootState.transactions = JSON.parse(JSON.stringify(emptyState))
    },
    async reconnected({ state, dispatch, rootState }) {
      // TODO: remove signedIn check when we support the option for setting a custom address for txs page
      if (state.loading && rootState.session.signedIn) {
        await dispatch(`getAllTxs`)
      }
    },
    async getAllTxs({ commit, dispatch, state, rootState }) {
      try {
        commit(`setHistoryLoading`, true)

        if (!rootState.connection.connected) return

        const bankTxs = await dispatch(`getTx`, `bank`)
        commit(`setBankTxs`, bankTxs)

        const stakingTxs = await dispatch(`getTx`, `staking`)
        commit(`setStakingTxs`, stakingTxs)

        const governanceTxs = await dispatch(`getTx`, `governance`)
        commit(`setGovernanceTxs`, governanceTxs)

        const distributionTxs = await dispatch(`getTx`, `distribution`)
        commit(`setDistributionTxs`, distributionTxs)

        const allTxs = bankTxs.concat(stakingTxs, governanceTxs, distributionTxs)

        await dispatch(`enrichTransactions`, allTxs)
        state.error = null
        commit(`setHistoryLoading`, false)
        state.loaded = true
      } catch (error) {
        Sentry.captureException(error)
        state.error = error
      }
    },
    async getTx({ rootState: { session: { address } } }, type) {
      let response
      const validatorAddress = address.replace(`cosmos`, `cosmosvaloper`)
      switch (type) {
        case `staking`:
          response = await node.getStakingTxs(address, validatorAddress)
          break
        case `governance`:
          response = await node.getGovernanceTxs(address)
          break
        case `distribution`:
          response = await node.getDistributionTxs(address, validatorAddress)
          break
        case `bank`:
          response = await node.txs(address)
          break
        default:
          throw new Error(`Unknown transaction type`)
      }
      const transactionsPlusType = response.map(t => ({ ...t, type }))
      return response ? uniqBy(transactionsPlusType, `txhash`) : []
    },
    async enrichTransactions({ dispatch }, transactions) {
      const blockHeights = new Set(
        transactions.map(({ height }) => parseInt(height))
      )
      await Promise.all(
        [...blockHeights].map(blockHeight =>
          dispatch(`queryTransactionTime`, { blockHeight })
        )
      )
    },
    async queryTransactionTime({ commit, dispatch }, { blockHeight }) {
      const blockMetaInfo = await dispatch(`queryBlockInfo`, blockHeight)
      commit(`setTransactionTime`, {
        blockHeight,
        time: blockMetaInfo.header.time
      })
    }
  }

  return {
    state,
    mutations,
    actions
  }
}
