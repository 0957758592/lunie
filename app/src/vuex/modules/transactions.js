import data from '../json/transactions.json'

const state = data

const mutations = {
}

export default {
  state,
  mutations
}
