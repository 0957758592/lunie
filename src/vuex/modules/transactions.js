import Vue from "vue"

// TODO simplify with one call

export default ({ node }) => {
  const emptyState = {
    loading: false,
    loaded: false,
    error: null,
    txs: []
  }

  const state = JSON.parse(JSON.stringify(emptyState))

  const mutations = {
    setHistoryLoading(state, loading) {
      Vue.set(state, `loading`, loading)
    },
    setTxs(state, txs) {
      state.txs = txs
    }
  }

  const actions = {
    resetSessionData({ rootState }) {
      // clear previous account state
      rootState.transactions = JSON.parse(JSON.stringify(emptyState))
    },
    async reconnected({ state, dispatch, rootState }) {
      // TODO: remove signedIn check when we support the option for setting a custom address for txs page
      if (state.loading && rootState.session.signedIn) {
        await dispatch(`getAllTxs`)
      }
    },
    async getAllTxs({ commit, state, rootState }) {
      try {
        commit(`setHistoryLoading`, true)

        if (!rootState.connection.connected) {
          return
        }

        const txs = await node.get.txs(rootState.session.address)
        commit("setTxs", txs)

        state.error = null
        commit(`setHistoryLoading`, false)
        state.loaded = true
      } catch (error) {
        state.error = error
      }
    },
    async enrichTransactions({ dispatch }, { transactions, txType }) {
      const enrichedTransactions = await Promise.all(
        transactions.map(async tx => {
          const blockMetaInfo = await dispatch(`queryBlockInfo`, tx.height)

          const enrichedTx = Object.assign({}, tx, {
            type: txType,
            time: new Date(blockMetaInfo.header.time).toISOString()
          })
          return enrichedTx
        })
      )
      return enrichedTransactions
    }
  }

  return {
    state,
    mutations,
    actions
  }
}
