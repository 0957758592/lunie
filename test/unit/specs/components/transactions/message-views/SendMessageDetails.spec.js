import { shallowMount } from "@vue/test-utils"
import { SendMessageDetails } from "src/components/transactions/message-view"
import { testTransactionObjects } from "test/unit/helpers/testValues"

describe(`SendMessageDetails`, () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SendMessageDetails, {
      propsData: {
        transaction: testTransactionObjects[0],
        sessionAddress: "cosmos1"
      }
    })
  })
  it(`MsgSend 1 has the expected html structure`, () => {
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`MsgSend 2 has the expected html structure`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[1] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`MsgSend 3 has the expected html structure`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[2] })
    expect(wrapper.element).toMatchSnapshot()
  })
  it(`MsgSend 4 has the expected html structure`, () => {
    wrapper.setProps({ transaction: testTransactionObjects[3] })
    expect(wrapper.element).toMatchSnapshot()
  })
})
