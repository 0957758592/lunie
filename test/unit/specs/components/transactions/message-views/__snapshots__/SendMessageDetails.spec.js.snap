// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendMessageDetails MsgSend 1 has the expected html structure 1`] = `
<div>
  <div
    class="tx__content__caption"
  >
    <p>
      
      Sent
      
      <b>
        0.2345
      </b>
       
      <span>
        ATOM
      </span>
    </p>
  </div>
   
  <div
    class="tx__content__information"
  >
    
      From 
      
    <bech32-stub
      address="cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkj"
    />
     to 
      
    <bech32-stub
      address="cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkj"
    />
     
    <span>
       - (Sent via Lunie)
    </span>
  </div>
</div>
`;

exports[`SendMessageDetails MsgSend 2 has the expected html structure 1`] = `
<div>
  <div
    class="tx__content__caption"
  >
    <p>
      
      Sent
      
      <b>
        0.2345
      </b>
       
      <span>
        ATOM
      </span>
    </p>
  </div>
   
  <div
    class="tx__content__information"
  >
    
      To 
      
    <bech32-stub
      address="cosmos1alkjdbasdlkhasldjnasldkjasndlkjansdlasndlkasdas"
    />
     
    <span>
       - (Sent via Lunie)
    </span>
  </div>
</div>
`;

exports[`SendMessageDetails MsgSend 3 has the expected html structure 1`] = `
<div>
  <div
    class="tx__content__caption"
  >
    <p>
      
      Sent
      
      <b>
        0.2345
      </b>
       
      <span>
        ATOM
      </span>
    </p>
  </div>
   
  <div
    class="tx__content__information"
  >
    To yourself! 
    <span>
       - (Sent via Lunie)
    </span>
  </div>
</div>
`;

exports[`SendMessageDetails MsgSend 4 has the expected html structure 1`] = `
<div>
  <div
    class="tx__content__caption"
  >
    <p>
      
      Received
      
      <b>
        0.2345
      </b>
       
      <span>
        ATOM
      </span>
    </p>
  </div>
   
  <div
    class="tx__content__information"
  >
    
      From 
      
    <bech32-stub
      address="cosmos11231212"
    />
     
    <span>
       - (Sent via Lunie)
    </span>
  </div>
</div>
`;
